### DISCUSSION TITLE
Is Gpu query and search slower than cpu?

### DISCUSSION THREAD

**Discussion by hjlee9182** [milvus-io/milvus#30691] [Q&A and General discussion] [answered]
In Milvues 2.3.9-gpu.
When just reading for collection with GPU_IVF_FLAT is slower than collection with IVF_FLAT.
How to improve search perforamnce..?

``` python
embedding = collection.query(
            output_fields=['embedding'],
            consistency_level='Bounded',
            expr=f'user_no == {user_no}',
        )[0]['embedding']

collection.search(**{
            'data': [embedding],
            'anns_field': 'embedding',
            'param': {
                'metric_type': 'IP',
                'params': {},
            },
            'output_fields': ['user_no'],
            'round_decimal': 4,
            'limit': 100,
        })
```

Collection
- 2M
- 512 emb size
- GPU_IVF_FLAT or IVF_FLAT

GPU_IVF_FLAT
<img width="616" alt="image" src="https://github.com/milvus-io/milvus/assets/46337218/c1810d96-587b-420b-8c19-30892970d085">
<img width="1075" alt="image" src="https://github.com/milvus-io/milvus/assets/46337218/30d0ebbd-15a4-46b4-9dd6-5ad252603645">

IVF_FLAT
<img width="502" alt="image" src="https://github.com/milvus-io/milvus/assets/46337218/562f6d8a-fad1-4412-a313-5d48e2544213">

**Comment by yhmo** [comment, marked as answer]
GPU index could be slower than CPU index if you only search a few vectors.
For example:
`collection.search(data = [vec_1], ...)`

GPU index is much more efficient than CPU index when searching large batch vectors.
For example:
`collection.search(data = [vec_1, vec_2, vec_3, ... vec_998, vec_999], ...)`

**Comment by xiaofan-luan** [comment]
> In Milvues 2.3.9-gpu. When just reading for collection with GPU_IVF_FLAT is slower than collection with IVF_FLAT. How to improve search perforamnce..?
> 
> ```python
> embedding = collection.query(
>             output_fields=['embedding'],
>             consistency_level='Bounded',
>             expr=f'user_no == {user_no}',
>         )[0]['embedding']
> 
> collection.search(**{
>             'data': [embedding],
>             'anns_field': 'embedding',
>             'param': {
>                 'metric_type': 'IP',
>                 'params': {},
>             },
>             'output_fields': ['user_no'],
>             'round_decimal': 4,
>             'limit': 100,
>         })
> ```
> 
> Collection
> 
> * 2M
> * 512 emb size
> * GPU_IVF_FLAT or IVF_FLAT
> 
> GPU_IVF_FLAT <img alt="image" width="616" src="https://private-user-images.githubusercontent.com/46337218/306176896-c1810d96-587b-420b-8c19-30892970d085.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MDg0MjA5OTgsIm5iZiI6MTcwODQyMDY5OCwicGF0aCI6Ii80NjMzNzIxOC8zMDYxNzY4OTYtYzE4MTBkOTYtNTg3Yi00MjBiLThjMTktMzA4OTI5NzBkMDg1LnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDAyMjAlMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQwMjIwVDA5MTgxOFomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPTM0YWZkMzFlZjcwMjBjMDljOGI1MmJlMjJmMjI2MDY3N2RkODZiNWVmMjI4MmVkMmUwYjkxYmFkNmQ5YWRhN2ImWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0JmFjdG9yX2lkPTAma2V5X2lkPTAmcmVwb19pZD0wIn0.SAjrVclfwzsqOPQ41FSEbQjDXLop_bLQ8MbieIy3OGw"> <img alt="image" width="1075" src="https://private-user-images.githubusercontent.com/46337218/306177192-30d0ebbd-15a4-46b4-9dd6-5ad252603645.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MDg0MjA5OTgsIm5iZiI6MTcwODQyMDY5OCwicGF0aCI6Ii80NjMzNzIxOC8zMDYxNzcxOTItMzBkMGViYmQtMTVhNC00NmI0LTlkZDYtNWFkMjUyNjAzNjQ1LnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDAyMjAlMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQwMjIwVDA5MTgxOFomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPTllZDFjZDdkNGM0YWQ5M2JkMjVmZmI1NmQzNGRkZGNmNDhkZmU2OTJiMzJlM2E1Mzk2YTA2YmIzOTRlOWM3ZTkmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0JmFjdG9yX2lkPTAma2V5X2lkPTAmcmVwb19pZD0wIn0.pIWDBFJ5bhi4GNLGdFFCSrQ1ucnz5yJLbnAhGOAJOm4">
> 
> IVF_FLAT <img alt="image" width="502" src="https://private-user-images.githubusercontent.com/46337218/306177073-562f6d8a-fad1-4412-a313-5d48e2544213.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MDg0MjA5OTgsIm5iZiI6MTcwODQyMDY5OCwicGF0aCI6Ii80NjMzNzIxOC8zMDYxNzcwNzMtNTYyZjZkOGEtZmFkMS00NDEyLWEzMTMtNWQ0OGUyNTQ0MjEzLnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDAyMjAlMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQwMjIwVDA5MTgxOFomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPTljYjg0Y2MyMTZjYjlmYzkxYzAxYjI2NDNhMTc1MGIzODAwOTM1MjM1YTg3NzBkMGU5MmJkYmJiOWRiYzU5ZGYmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0JmFjdG9yX2lkPTAma2V5X2lkPTAmcmVwb19pZD0wIn0.9b1hOkTkixLhqh_z5NoBX3UzLA-P1ifP50DSsQkVn_A">

Please help patience and wait for Milvus 2.4.
We have a special GPU index which is 10 times faster than GPU IVF FLAT

**Comment by xiaofan-luan** [comment]
Here is some performance result, stay tuned

<img width="827" alt="image" src="https://github.com/milvus-io/milvus/assets/83447078/5a08a1f0-f6d3-4e57-9f34-d6c46749154c">
